language/stdlib/modules/libra_account.mvir

```diff
@@ -57,22 +76,93 @@ module LibraAccount {
         payer: address,
     }
 
+    // Events
+    // Derive a fresh unique id by using sender's HandleIdGenerator. The generated bytearray is indeed unique because it
+    // was derived from the hash(sender's HandleIdGenerator || sender_address). This module guarantees that the
+    // HandleIdGenerator is only going to be monotonically increased and there's no way to revert it or destroy it. Thus
+    // such counter is going to give distinct value for each of the new event stream under each sender. And since we
+    // hash it with the sender's address, the result is guaranteed to be globally unique.
+    fresh_guid(counter: &mut Self.HandleIdGenerator): bytearray{
+        let count: &mut u64;
+        let sender_bytes: bytearray;
+        let count_bytes: bytearray;
+        let preimage: bytearray;
+
+        count = &mut move(counter).counter;
+
+        sender_bytes = AddressUtil.address_to_bytes(get_txn_sender());
+
+        count_bytes = U64Util.u64_to_bytes(*copy(count));
+        *move(count) = *copy(count) + 1;
+
+        // HandleIdGenerator goes first just in case we want to extend address in the future.
+        preimage = BytearrayUtil.bytearray_concat(move(count_bytes), move(sender_bytes));
```

Comment:
Assuming that request ~= transaction :)

We discussed that option, but it's problematic for a few reasons. The big problem is: what code has the authority/permission to increment the counter? There are two obvious options:
(1) Only the code that creates event handles can increment it. In that case, the counter isn't useful to anyone else, and we're adding "transaction-global" state that only the event handle code can access.
(2) Anyone can increment the counter. In that case, it can't be used to guarantee uniqueness because you can forget to increment, plus we've added global mutable state that anyone can touch. If everyone can touch it, no one can rely on it (except that it will always be increasing).

We decided the VM/prologue complexity and potential confusion of those choices wasn't worth saving a few bytes. 

Another alternative we discussed was using `get_gas_remaining()`, which would accomplish the same thing without adding transaction-global state, but has a variety of different problems.

Reply:
I was thinking of something like a function `next_per_txn_id()`. The Nth call to this function returns the integer N. The guarantee of this function would be that the tuple (sender_address, txn_sequence_number, next_per_txn_id()) is unique. Even if there are multiple callers of this function, they can't break the use case (since the function both accesses and increments).

Anyways -- glad the approach was discussed. I'll leave it to you guys to prioritize the optimization here. Curious how you guys think about the cost of per-account data. This kind of thing obviously isn't worth it for saving a few bytes, but what about a few bytes per account? Also wondering if other people may end up needing a similar GUID

Reply:
> I was thinking of something like a function next_per_txn_id(). The Nth call to this function returns the integer N. The guarantee of this function would be that the tuple (sender_address, txn_sequence_number, next_per_txn_id()) is unique. Even if there are multiple callers of this function, they can't break the use case (since the function both accesses and increments).

This what option (2) describes and it would certainly work. You would do it by adding `PerTxnCounter.initialize()` to the tx prologue  and `PerTxnCounter.clear()` to the tx epilogue. But our bar for adding new logic there is pretty high; we try to limit it to stuff that every transaction will need to do (and have succeeded so far).

> Curious how you guys think about the cost of per-account data. This kind of thing obviously isn't worth it for saving a few bytes, but what about a few bytes per account?

My philosophy here is: ruthlessly prioritize the simplicity and predictability of the language, gas behavior, virtual machine, and core modules in our first cut, then optimize for space once the semantic stuff stabilizes. Otherwise, we might do something complex for a space savings that ends up not being important, or building a new feature on top of the complex thing might end up costing more space than building a new feature on top of the simple thing.

I fully expect that eventually there will be a reckoning on space at some point and we will need to make tough calls on things like "add `PerTxnCounter` logic to the prologue/epilogue to make accounts and events smaller".

> Also wondering if other people may end up needing a similar GUID

This is a good question. In almost every case, I think Move programmers should define a new type when they need something that's globally unique. This is safer, more space-efficient, and more flexible than trying to create a unique piece of data. Would have loved to do `EventKey`s in this way, but the type system can't quite make it work...

In some cases, you might want to do something similar to a GUID if you want to create an unbounded number of unique instances of a particular type (e.g., serial numbers for some NFT). You would want to create your own module-local, per-account GUID rather than reuse some shared global GUID. Otherwise, a griefer can exhaust the supply of ID's by creating a new one in a loop (which wouldn't take that long/that much gas for a u64, and bigger counters would take up more gas). This is also why we chose to make our GUID per-account rather than global...

I do think it would make sense to (eventually) extract the GUID generation logic into some generic type defined outside of `LibraAccount` to allow others to reuse it.